File: 1302.层数最深叶子节点的和.cpp
/*
 * @lc app=leetcode.cn id=1302 lang=cpp
 *
 * [1302] 层数最深叶子节点的和
 */
#include <queue>
struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode() : val(0), left(nullptr), right(nullptr) {}
  TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
  TreeNode(int x, TreeNode *left, TreeNode *right)
      : val(x), left(left), right(right) {}
};
// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
  /**
   * @brief
   * 39/39 cases passed (88 ms)
   * Your runtime beats 65.02 % of cpp submissions
   * Your memory usage beats 27.3 % of cpp submissions (60.4 MB)
   * @param root
   * @return int
   */
  int deepestLeavesSum(TreeNode *root) {
    std::queue<TreeNode *> bfs;
    bfs.push(root);
    for (; !bfs.empty();) {
      auto num = bfs.size();
      int ans = 0;
      for (; num--;) {
        auto fr = bfs.front();
        bfs.pop();
        ans += fr->val;
        if (fr->left) {
          bfs.push(fr->left);
        }
        if (fr->right) {
          bfs.push(fr->right);
        }
      }
      if (bfs.size() == 0) {
        return ans;
      }
    }
    return 0;
  }
};
// @lc code=end

File: 1302.层数最深叶子节点的和.java
/*
 * @lc app=leetcode.cn id=1302 lang=java
 *
 * [1302] 层数最深叶子节点的和
 */
import java.util.concurrent.ConcurrentLinkedQueue;

class TreeNode {

  int val;
  TreeNode left;
  TreeNode right;

  TreeNode() {}

  TreeNode(int val) {
    this.val = val;
  }

  TreeNode(int val, TreeNode left, TreeNode right) {
    this.val = val;
    this.left = left;
    this.right = right;
  }
}

// @lc code=start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {

  /**
   * 39/39 cases passed (11 ms)
   * Your runtime beats 6.95 % of java submissions
   * Your memory usage beats 89.11 % of java submissions (43.4 MB)
   * @param root
   * @return
   */
  public int deepestLeavesSum(TreeNode root) {
    var bfs = new ConcurrentLinkedQueue<TreeNode>();
    bfs.add(root);
    for (; !bfs.isEmpty();) {
      var size = bfs.size();
      var ans = 0;
      for (; size-- > 0;) {
        var fr = bfs.poll();
        ans += fr.val;
        if (fr.left != null) {
          bfs.add(fr.left);
        }
        if (fr.right != null) {
          bfs.add(fr.right);
        }
      }
      if (bfs.size() == 0) {
        return ans;
      }
    }
    return 0;
  }
}
// @lc code=end
